### DISCUSSION TITLE
IFM device discovery and change of the IP

### DISCUSSION THREAD

**Discussion by matijazigic** [ifm/ifm3d#468] [Q&A]
Hello everyone,

I've been trying for quite some time to find a way to connect to the VPU and cameras programmatically or from the command line, but without success. For my company, it is important to have this functionality because we are running three cameras on a vehicle, and a fast configuration and calibration procedure is essential for our end-of-line process.

As mentioned in other [topics](https://github.com/ifm/ifm3d/discussions/433) and [here](https://github.com/ifm/ifm3d/issues/194), there is an auto-discovery option and the ability to change the IP address when the VPU device is found. This would be extremely useful for us to integrate into our workflow. I am able to do this with the ifmVisionAssistant, but we prefer not to use two separate software solutions. We want the process to be as fast as possible, and some of our colleagues do not have access to ifmVisionAssistant.

Below are screenshots showing what I get with the ifm3d discover option and with the VisionAssistant:

![ifm1](https://github.com/user-attachments/assets/edd26768-c79a-449b-b94a-144df8c215df)

![ifm2](https://github.com/user-attachments/assets/2e5cbd6a-808d-4598-8801-ea387021c748)

Additionally, here is what I get when running the Python script for ifm discovery mentioned earlier in the discussions. I encounter the same error on both Windows and Ubuntu. Am I doing something wrong?

![ifm3](https://github.com/user-attachments/assets/0c0f6eb1-410f-4429-9390-8624c499050b)

Do you have any other options for identifying the IP and changing it? The user will not be on the same subnet, so we need to replicate what is done in the ifmVisionAssistant (image 2) — specifically, retrieving the device and programmatically changing its IP address.

Perhaps @graugans or @lola-masson could help?

**Comment by lola-masson** [comment]
Hi @matijazigic,

Looking at your first screenshot, it looks like the device is actually found, only the device type cannot be identified:
![image](https://github.com/user-attachments/assets/77fd1683-f23f-442e-85c7-b068f6632914)

Regarding setting a temporary IP address using the CLI, this is a feature we are currently implementing, so it should be available shortly. Until then, you would need to set your computer in the same subnet.

**Reply by matijazigic** [reply]
Thank you @lola-masson, do you know when it would be available?

**Reply by lola-masson** [reply]
It should be on main branch next week, but we probably won't have it in an official release quite yet, thought that is not yet decided. You are welcome to give it a try building from source when it is available, though.

**Reply by matijazigic** [reply]
Hello @lola-masson, is there any updates maybe?

**Reply by lola-masson** [reply]
The feature was added to main today: 9b13f800023c680ae17a0ef1c31b061e28f1d981 :partying_face: 
To test this, you should build the library from source. Let us know if you need any further assistance.

**Reply by matijazigic** [reply]
Is there any news when it would be official? Is it going to be available via ifm3dpy?

**Reply by lola-masson** [reply]
We are planning on creating an official release in January, which will come with Python bindings. The Python bindings will be available earlier on the main branch, so you can keep an eye out if you wanted to test them before the official release.

**Reply by matijazigic** [reply]
Hello @lola-masson, is official update going to be out soon?

**Reply by graugans** [reply]
Maybe a bit late in the game, there is this archived version of the Python script I wrote ages ago: https://github.com/ifm/python-ipv4-discovery/blob/master/discovery.py as far as I do remember there was an issue with sending request on a distinct interface without root access on Linux.

**Reply by matijazigic** [reply]
Hello @graugans yes, I tried that one over a year ago. We had problems while using that script with the detection of the devices, and also there is no possibility to change the IP once the device is discovered.

**Reply by lola-masson** [reply]
Hi @matijazigic, we are planning on doing a pre-release at the end of the week, followed by our tests, which means we should have an official release within the next few weeks.

**Reply by graugans** [reply]
> Hello @graugans yes, I tried that one over a year ago. We had problems while using that script with the detection of the devices, and also there is no possibility to change the IP once the device is discovered.

I am not sure if I fully understand your sentence __there is no possibility to change the IP once the device is discovered__ Within the script the ip can be changed: https://github.com/ifm/python-ipv4-discovery/blob/master/discovery.py#L114

**Comment by matijazigic** [comment]
I fixed some problems with the print statements inside the script but still there is an error:

`[sudo] password for hrjhmz00: 
IP: 192.168.0.69
interface: eth0
Traceback (most recent call last):
  File "/home/code_sandbox/3D/o3r_framework/discovery.py", line 75, in <module>
    s.setsockopt(socket.SOL_SOCKET, 25, args.interface) # 25 is the magic number to use a specific interface 'SO_BINDTODEVICE'
TypeError: a bytes-like object is required, not 'str'`

I can try to fix it, but exact same errors I got before. Where can I find updated script? Thanks for your help.

**Comment by matijazigic** [comment]
I fixed some problems with the print statements inside the script but still there is an error:

`[sudo] password for hrjhmz00: 
IP: 192.168.0.69
interface: eth0
Traceback (most recent call last):
  File "/home/code_sandbox/3D/o3r_framework/discovery.py", line 75, in <module>
    s.setsockopt(socket.SOL_SOCKET, 25, args.interface) # 25 is the magic number to use a specific interface 'SO_BINDTODEVICE'
TypeError: a bytes-like object is required, not 'str'`

I can try to fix it, but it is same errors I got before. Where can I find updated script?

**Reply by matijazigic** [reply]
I have fixed that error by using encode:
`s.setsockopt(socket.SOL_SOCKET, 25, args.interface.encode('utf-8'))`

After that I'm getting:

 `IP: 192.168.0.69
interface: eth0
Trying to find server...
Could not find ipv4 discovery server timed out`

Script cannot find the device.

**Reply by matijazigic** [reply]
I can connect to the IFM VPU via IFM VA and ping it without problems.
```
ping 192.168.0.69                                                                                                                                                                                                                                          5s Py venv_python3.10 13:49:08
PING 192.168.0.69 (192.168.0.69) 56(84) bytes of data.
64 bytes from 192.168.0.69: icmp_seq=1 ttl=63 time=0.482 ms
64 bytes from 192.168.0.69: icmp_seq=2 ttl=63 time=0.382 ms
64 bytes from 192.168.0.69: icmp_seq=3 ttl=63 time=0.700 ms
```

**Reply by graugans** [reply]
I guess this is needed to fix the Python3 transistion:

```diff
diff --git a/discovery.py b/discovery.py
index aaa77ebe76a030246f2f75e1560b681b3cfe46d7..3a56aec08296f0b81ed2fbf9926a493cb9d96b4c 100755
--- a/discovery.py
+++ b/discovery.py
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 
 # Copyright (C) ifm electronic gmbh - All Rights Reserved
 # Unauthorized copying of this file, via any medium is strictly prohibited
@@ -20,7 +20,7 @@
 # Import of needed python modules
 import socket
 import threading
-from SocketServer import UDPServer, BaseRequestHandler
+from socketserver import UDPServer, BaseRequestHandler
 import struct
 import argparse
 import sys
@@ -40,7 +40,7 @@ PORT = 3321
 
 def change_ip(s,ip,mac):
     """Helper function to change the IP address"""
-    print('Change IP of the device to: {} with mac: {}'.format(ip,":".join("{:02x}".format(ord(c)) for c in device_mac)))
+    print('Change IP of the device to: {} with mac: {}'.format(ip,":".join("{:02x}".format(c) for c in device_mac)))
     set_ip = bytearray([0]*20) # create a zero byte buffer
     set_ip[0:4] = [0x10,0x20,0xef,0xce] # inject the magic bytes
     set_ip[4:6] = struct.pack('>H',PORT)
@@ -74,7 +74,7 @@ if __name__ == "__main__":
         s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, True)
         # TODO: use a way to detect all interfaces to send the broadcasts
         # here we use a hardcoded interface as an example
-        s.setsockopt(socket.SOL_SOCKET, 25, args.interface) # 25 is the magic number to use a specific interface 'SO_BINDTODEVICE'
+        s.setsockopt(socket.SOL_SOCKET, 25, str(args.interface).encode('utf-8')) # 25 is the magic number to use a specific interface 'SO_BINDTODEVICE'
         s.bind(('', PORT))
         s.settimeout(5)
     except IOError:
@@ -106,9 +106,10 @@ if __name__ == "__main__":
             print('Device-ID:          0x{:04x}'.format(device_id))
             mac=response[32:38]
             device_mac = response[32:38]
-            print('Device-MAC:         {}'.format(":".join("{:02x}".format(ord(c)) for c in device_mac)))
+            # print('Device-MAC:         {}'.format(":".join("{:02x}".format(ord( chr(c ))) for c in device_mac)))
+            print('Device-MAC:         {}'.format(":".join("{:02x}".format( c ) for c in device_mac)))
             device_flags = struct.unpack('>H',response[38:40])[0]
-            print('Device-Flags:       0x{:04x}'.format(device_id))
+            print('Device-Flags:       0x{:04x}'.format(device_flags))
             device_hostname = response[40:104]
             print('Device-Hostname:    {}'.format(device_hostname))
             device_devicename = response[104:]
````

**Reply by graugans** [reply]
What firmware Version and which VPU you are using? Are you running the script inside a VM or an environment like WSL2?

**Reply by matijazigic** [reply]
VPU: M04239
Firmware version: 
![image](https://github.com/user-attachments/assets/91d6cb40-0221-426e-9481-68dd5b98b800)

I'm running script in the WSL2 environment.

**Reply by matijazigic** [reply]
I tried with all changes you introduced, problem still exists.

**Reply by graugans** [reply]
Okay, the script does execute without a Python Error but still you are not able to change the IP right?

**Reply by graugans** [reply]
> VPU: M04239 Firmware version: ![image](https://private-user-images.githubusercontent.com/8402071/412433772-91d6cb40-0221-426e-9481-68dd5b98b800.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzkzNjcwMDgsIm5iZiI6MTczOTM2NjcwOCwicGF0aCI6Ii84NDAyMDcxLzQxMjQzMzc3Mi05MWQ2Y2I0MC0wMjIxLTQyNmUtOTQ4MS02OGRkNWI5OGI4MDAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMTJUMTMyNTA4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZjZhMmVmNTYwNmZjYjBkM2E1YTc2NjQ2NmYwZDJjZDRhOWZkNjc1NjM4ZWFhZDc5NzExNjQzYzIxOWU5ZmEyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.sTIgMpP0uJbCQhO5_2BXjOtTGdFCiAyrCkD_jUdiEu8)
> 
> I'm running script in the WSL2 environment.

I am not sure if the broadcast will work in an WSL2 Environment. Do you have an RaspBerry PI running Linux or maybe a second VPU available to test this?

**Reply by matijazigic** [reply]
Is it possible to run script from windows machine?

**Reply by matijazigic** [reply]
I got response via Windows; I made couple changes on the script. 
```
PS D:\SW\AGVEP-o3r_tools> python discover2.py -i 192.168.0.143
No IP address specified. Just reading responses.
Trying to find server...
Magic matched, sender: ('0.0.0.0', 3321)
Device IPv4 addr:   192.168.0.69
Gateway IPv4 addr:  192.168.0.201
IPv4 subnetmask:    255.255.255.0
XML-RPC port:       8080
Vendor-ID:          0x0001
Device-ID:          0x0302
Device-MAC:         48:b0:2d:54:d7:fd
Device-Flags:       0x0000
Device-Hostname:    ovp80x-54-d7-fd
Device-Name:
```
